### PR TITLE
bluetooth: Remove useless LE Set Data Length after Connected

### PR DESCRIPTION
**Pull request by rugeGerritsen**
Calling LE Write Suggested Default Data Length command is sufficient
to ensure that the LL will auto-initiate the DLE procedure. Performing
a second procedure is unecessary.

As some controllers don't follow the spec recommendation, a controller quirk is added

In Core v5.2, Vol 4, Part E, Section 7.8.35:
_The HCI_LE_Write_Suggested_Default_Data_Length command allows the
Host to specify its suggested values for the Controller's maximum
transmission number of payload octets and maximum packet
transmission time for packets containing LL Data PDUs to be used
for new connections. The Controller may use smaller or larger values
for connInitialMaxTxOctets and connInitialMaxTxTime based on local
information._

In short, this command sets connInitialMaxTxOctets
and connInitialMaxTxTime.

In Core v5.2, Vol 6, Part B, Section 4.5.10:
_For a new connection: connMaxTxOctets shall be set to
connInitialMaxTxOctets and connMaxRxOctets shall be chosen by the
Controller. If either value is not 27 then the Controller should
initiate the Data Length Update Procedure at the earliest practical
opportunity._

Therefore, calling LE Write Suggested Default Data Length command
is sufficient to ensure the DLE being performed.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

### CODE REVIEW

**Comment by rugeGerritsen** [comment]
@cvinayak . You may be interested in this as well. The Zephyr controller does not auto-initiate the DLE procedure as required by spec.

**Comment by cvinayak** [comment]
@rugeGerritsen its a "should" not a "shall".

**Comment by joerchan** [comment]
@rugeGerritsen We need to remove the conn state flag `BT_CONN_AUTO_DATA_LEN_COMPLETE` it will prevent the user controlled data len procedure to be called now.

**Comment by rugeGerritsen** [comment]
> @rugeGerritsen its a "should" not a "shall".

Thanks for pointing out. In either case, it would be nice if the host assumes the behavior recommended by the specification.

**Comment by rugeGerritsen** [comment]
> @rugeGerritsen We need to remove the conn state flag `BT_CONN_AUTO_DATA_LEN_COMPLETE` it will prevent the user controlled data len procedure to be called now.

Done

**Comment by joerchan** [comment]
@carlescufi @rugeGerritsen @cvinayak Shall (should?) we come to an agreement here?
The question is if we would want to move this handling to zephyr controller or not.
The host should assume that the HCI controller uses the recommended practice, that the LE Write Default DLE/PHY would auto-initiate the procedure in the controller. 
Maybe we could put sending the LE Set Data Len as an HCI driver quirk?

**Comment by carlescufi** [comment]
> @carlescufi @rugeGerritsen @cvinayak Shall (should?) we come to an agreement here?
> The question is if we would want to move this handling to zephyr controller or not.
> The host should assume that the HCI controller uses the recommended practice, that the LE Write Default DLE/PHY would auto-initiate the procedure in the controller.
> Maybe we could put sending the LE Set Data Len as an HCI driver quirk?

I am in favor of delegating this to the controller, although it will require a bit of an effort. In the meantime I think adding a quirk (or a Kconfig option) seems like the path forward in order to be compatible with both kinds of controllers.
I have also added @kruithofa to the reviewer list so that he can discuss with @cvinayak how best to implement this in the controller.

**Comment by rugeGerritsen** [comment]
@cvinayak , @kruithofa , do you have input to this issue?

**Comment by joerchan** [comment]
@rugeGerritsen 

> I agree that it would be best to implement this in the controller. I believe it is quite simple to implement.
> 
> We still may need the configurability in the host. What about the following proposal: Split the CONFIG_BT_AUTO_DATA_LEN_UPDATE to CONFIG_BT_SET_DEFAULT_DATA_LEN and CONFIG_BT_DATE_LEN_UPDATE_ON_CONN_COMPLETE. We can then add overlays or something similar for the different controllers. A similar approach can be done for the PHY procedure.

I don't think we should have this as Kconfig, as it is more of a strangeness than a feature. I think either BT_QUIRK_ (hci_driver.h) or maybe just wrap that code in ifdef LL_SW_SPLIT until we have it fixed in the controller.

**Comment by rugeGerritsen** [comment]
> I don't think we should have this as Kconfig, as it is more of a strangeness than a feature. I think either BT_QUIRK_ (hci_driver.h) or maybe just wrap that code in ifdef LL_SW_SPLIT until we have it fixed in the controller.

Wrapping it in ifdef LL_SW_SPLIT won't work in multicore systems. We therefore need to implement it either as a quick or in the controller.

**Comment by joerchan** [comment]
I would add a quirk, and then remove the whole the quirk again when it is implemented by the controller.

File: subsys/bluetooth/host/hci_core.c [hunk 1]
```diff
@@ -1941,15 +1941,21 @@ static void conn_auto_initiate(struct bt_conn *conn)
 	}
 
 	if (IS_ENABLED(CONFIG_BT_AUTO_DATA_LEN_UPDATE) &&
-	    BT_FEAT_LE_DLE(bt_dev.le.features)) {
-		uint16_t tx_octets, tx_time;
+		BT_FEAT_LE_DLE(bt_dev.le.features)) {
```
Comment (joerchan) [inline subsys/bluetooth/host/hci_core.c:1944]: align with rest of if-statement.

File: subsys/bluetooth/host/hci_core.c [hunk 2]
```diff
@@ -2320,7 +2326,8 @@ static void le_data_len_change(struct net_buf *buf)
 	       max_tx_time, max_rx_octets, max_rx_time);
 
 #if defined(CONFIG_BT_USER_DATA_LEN_UPDATE)
-	if (IS_ENABLED(CONFIG_BT_AUTO_DATA_LEN_UPDATE)) {
+	if (IS_ENABLED(CONFIG_BT_AUTO_DATA_LEN_UPDATE) &&
+		(bt_dev.drv->quirks & BT_QUIRK_NO_AUTO_DLE)) {
```
Comment (joerchan) [inline subsys/bluetooth/host/hci_core.c:2330]: align with rest of if-statement

File: subsys/bluetooth/host/hci_core.c [hunk 3]
```diff
@@ -1941,15 +1941,21 @@ static void conn_auto_initiate(struct bt_conn *conn)
...
+			if (!err) {
+				err = bt_le_set_data_len(conn, tx_octets, tx_time);
+				if (err) {
+					BT_ERR("Failed to set data len (%d)", err);
+				}
 			}
+		} else {
+			/* No need to auto-initiate DLE procedure.
+			 * It is done by the controller. */
+			atomic_set_bit(conn->flags, BT_CONN_AUTO_DATA_LEN_COMPLETE);
```
Comment (joerchan) [inline subsys/bluetooth/host/hci_core.c:1958]: Wouldn't it be enough to only set this in the le_data_len_change?
Now the application can start the user initiated data length update procedure while the controller initiated data length update might be running.

File: subsys/bluetooth/host/hci_core.c [hunk 4]
```diff
@@ -5688,28 +5695,34 @@ static int le_init(void)
 	if (IS_ENABLED(CONFIG_BT_CONN) &&
 	    IS_ENABLED(CONFIG_BT_DATA_LEN_UPDATE) &&
 	    BT_FEAT_LE_DLE(bt_dev.le.features)) {
-		struct bt_hci_cp_le_write_default_data_len *cp;
-		uint16_t tx_octets, tx_time;
+		if ((IS_ENABLED(CONFIG_BT_AUTO_DATA_LEN_UPDATE) &&
+			!(bt_dev.drv->quirks & BT_QUIRK_NO_AUTO_DLE))) {
```
Comment (joerchan) [inline subsys/bluetooth/host/hci_core.c:5699]: align with the rest of the if-statement.

File: subsys/bluetooth/host/hci_core.c [hunk 5]
```diff
@@ -5688,28 +5695,34 @@ static int le_init(void)
...
-					   buf, NULL);
-		if (err) {
-			return err;
+			err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_WRITE_DEFAULT_DATA_LEN,
+						   buf, NULL);
+			if (err) {
+				return err;
+			}
+		} else {
+			/* No need to set default data length.
```
Comment (joerchan) [inline subsys/bluetooth/host/hci_core.c:5724]: Comment is a misleading if CONFIG_BT_AUTO_DATA_LEN_UPDATE is disabled

**Review by joerchan — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/hci_core.c [hunk 6]
```diff
@@ -170,7 +170,7 @@ NET_BUF_POOL_FIXED_DEFINE(hci_rx_pool, CONFIG_BT_RX_BUF_COUNT,
 			  BT_BUF_RX_SIZE, NULL);
 
 #if defined(CONFIG_BT_CONN)
-#define NUM_COMLETE_EVENT_SIZE BT_BUF_SIZE(                            \
+#define NUM_COMLETE_EVENT_SIZE BT_BUF_SIZE(\
```
Comment (joerchan) [inline subsys/bluetooth/host/hci_core.c:173]: accidental change here?

**Review by joerchan — APPROVED** [review]
Looks good, thank you.

**Comment by cvinayak** [comment]
> 
> 
> @cvinayak , @kruithofa , do you have input to this issue?

The auto-initiation in the current controller LLCP, not going to invest. In the new LLCP, yes, can be added as select-able feature.

**Comment by joerchan** [comment]
>  In the new LLCP, yes, can be added as select-able feature.

I think it is better to just make that only behavior, otherwise what is the point of the set default command?

**Comment by cvinayak** [comment]
> 
> 
> > In the new LLCP, yes, can be added as select-able feature.
> 
> I think it is better to just make that only behavior, otherwise what is the point of the set default command?

The values from the set default length command will get used in the length response when the auto-initiation is disabled. As the host does not auto-initiate ATT MTU request, or would synchronize with link layer as to when length request is sent, it should be at the applications discretion to optimally use these procedure. 

Both data length and phy update procedures where auto-initiated by default in host, and it has been now made as Kconfig configurable.

**Comment by joerchan** [comment]
@cvinayak Ok. Do you have any comments to this PR and the way the host now uses these commands?

File: include/drivers/bluetooth/hci_driver.h [hunk 7]
```diff
@@ -30,6 +30,11 @@ extern "C" {
 enum {
 	/* The host should never send HCI_Reset */
 	BT_QUIRK_NO_RESET = BIT(0),
+	/* The controller does not auto-initiate a DLE procedure when the
+	 * initial connection data length parameters are not equal to the
+	 * default data length parameters. Therefore the host should initiate
+	 * the DLE procedure after connection establishment. */
+	BT_QUIRK_NO_AUTO_DLE = BIT(1),
 };
```
Comment (cvinayak) [inline include/drivers/bluetooth/hci_driver.h:39]:
```suggestion
#define IS_BT_QUIRK_NO_AUTO_DLE(bt_dev) ((bt_dev)->drv.quirks & BT_QUIRK_NO_AUTO_DLE)
```

File: subsys/bluetooth/host/hci_core.c [hunk 8]
```diff
@@ -1942,14 +1942,21 @@ static void conn_auto_initiate(struct bt_conn *conn)
 
 	if (IS_ENABLED(CONFIG_BT_AUTO_DATA_LEN_UPDATE) &&
 	    BT_FEAT_LE_DLE(bt_dev.le.features)) {
-		uint16_t tx_octets, tx_time;
+		if (bt_dev.drv->quirks & BT_QUIRK_NO_AUTO_DLE) {
```
Comment (cvinayak) [inline subsys/bluetooth/host/hci_core.c:1945]:
```suggestion
		if (IS_BT_QUIRK_NO_AUTO_DLE(&bt_dev)) {
```

**Review by cvinayak — CHANGES_REQUESTED** [review]
Minor nitpick on use of a check macro; Otherwise, changes look good to me.

**Review by cvinayak — APPROVED** [review]

**Comment by rugeGerritsen** [comment]
Rebased to get rid of the conflict in hci_driver.h

**Comment by joerchan** [comment]
@jhedberg Can you have a look?

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
Ops, I thought this was reasonable so I merged, but I failed to see that @joerchan had asked @jhedberg to take a look. @jhedberg sorry about that, let  me know if you have any objections.
